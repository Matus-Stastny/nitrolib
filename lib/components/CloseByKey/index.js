"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var React = _interopRequireWildcard(require("react"));

var ESCAPE = "Escape";

var CloseByKey =
/*#__PURE__*/
function (_React$PureComponent) {
  (0, _inherits2.default)(CloseByKey, _React$PureComponent);

  function CloseByKey() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2.default)(this, CloseByKey);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2.default)(this, (_getPrototypeOf2 = (0, _getPrototypeOf3.default)(CloseByKey)).call.apply(_getPrototypeOf2, [this].concat(args)));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)((0, _assertThisInitialized2.default)(_this)), "handleKeyDown", function (ev) {
      ev.stopPropagation();
      var _this$props = _this.props,
          closeKey = _this$props.closeKey,
          onClose = _this$props.onClose;

      if (onClose && ev.key === closeKey) {
        onClose(ev);
      }
    });
    return _this;
  }

  (0, _createClass2.default)(CloseByKey, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      var _document = document,
          body = _document.body;

      if (body) {
        body.addEventListener("keydown", this.handleKeyDown);
      }
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      var _document2 = document,
          body = _document2.body;

      if (body) {
        body.removeEventListener("keydown", this.handleKeyDown);
      }
    }
  }, {
    key: "render",
    value: function render() {
      var children = this.props.children;
      return children;
    }
  }]);
  return CloseByKey;
}(React.PureComponent);

exports.default = CloseByKey;
(0, _defineProperty2.default)(CloseByKey, "defaultProps", {
  closeKey: ESCAPE
});