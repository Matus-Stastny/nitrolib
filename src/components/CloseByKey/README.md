# CloseByKey

Fires a callback whenever a user presses the close button (_Escape_ by default).
