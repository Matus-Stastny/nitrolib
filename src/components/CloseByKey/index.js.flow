// @flow strict
import * as React from "react";

type Props = {|
  onClose: (ev: KeyboardEvent) => void,
  children: React.Node,
  // defaulted
  closeKey?: string, // Escape
|};

declare export default React.ComponentType<Props>;
