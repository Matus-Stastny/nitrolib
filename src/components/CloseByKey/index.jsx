// @flow strict
import * as React from "react";

const ESCAPE = "Escape";

type Props = {|
  closeKey: string,
  children: React.Node,
  onClose: (event?: KeyboardEvent) => void,
|};

export default class CloseByKey extends React.PureComponent<Props> {
  static defaultProps = {
    closeKey: ESCAPE,
  };

  componentDidMount() {
    const { body } = document;

    if (body) {
      body.focus();
      body.addEventListener("keydown", this.handleKeyDown);
    }
  }

  componentWillUnmount() {
    const { body } = document;

    if (body) {
      body.removeEventListener("keydown", this.handleKeyDown);
    }
  }

  handleKeyDown = (event: KeyboardEvent) => {
    event.stopPropagation();

    const { closeKey, onClose } = this.props;

    if (onClose && event.key === closeKey) {
      onClose(event);
    }
  };

  render() {
    const { children } = this.props;

    return children;
  }
}
