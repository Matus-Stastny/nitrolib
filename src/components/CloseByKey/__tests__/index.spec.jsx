// @flow strict
import * as React from "react";
import { shallow } from "enzyme";

import CloseByKey from "..";

describe("#CloseByKey", () => {
  test("render", () => {
    const wrapper = shallow(
      <CloseByKey onClose={jest.fn()}>
        <h1>kek</h1>
      </CloseByKey>,
    );

    expect(wrapper).toMatchSnapshot();
  });

  test("close on specified key", () => {
    const map = {};
    // $FlowExpected: TODO describe
    document.body.addEventListener = jest.fn((event, cb) => {
      map[event] = cb;
    });

    const onClose = jest.fn();
    shallow(
      <CloseByKey onClose={onClose} closeKey="Home">
        <h1>kek</h1>
      </CloseByKey>,
    );

    const stopPropagation = jest.fn();
    map.keydown({ key: "Home", stopPropagation });

    expect(stopPropagation).toBeCalled();
    expect(onClose).toBeCalled();
  });

  test("not close on other keys", () => {
    const map = {};
    // $FlowExpected: TODO describe
    document.body.addEventListener = jest.fn((event, cb) => {
      map[event] = cb;
    });

    const onClose = jest.fn();
    shallow(
      <CloseByKey onClose={onClose} closeKey="Home">
        <h1>kek</h1>
      </CloseByKey>,
    );

    const stopPropagation = jest.fn();
    map.keydown({ key: "Delete", stopPropagation });

    expect(stopPropagation).toBeCalled();
    expect(onClose).not.toBeCalled();
  });

  test("stop listening on unmount", () => {
    const map = {};
    // $FlowExpected: TODO describe
    document.body.addEventListener = jest.fn((event, cb) => {
      map[event] = cb;
    });
    const removeListener = jest.fn();
    // $FlowExpected: TODO describe
    document.body.removeEventListener = removeListener;

    const onClose = jest.fn();
    const wrapper = shallow(
      <CloseByKey onClose={onClose}>
        <h1>kek</h1>
      </CloseByKey>,
    );

    wrapper.instance().componentWillUnmount();

    expect(removeListener).toBeCalledWith("keydown", map.keydown);
  });
});
