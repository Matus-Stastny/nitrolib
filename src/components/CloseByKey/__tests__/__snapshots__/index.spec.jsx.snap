// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#CloseByKey render 1`] = `
<h1>
  kek
</h1>
`;
